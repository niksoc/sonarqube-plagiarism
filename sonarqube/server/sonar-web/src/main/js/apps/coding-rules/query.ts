/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  cleanQuery,
  parseAsArray,
  parseAsDate,
  parseAsOptionalBoolean,
  parseAsOptionalString,
  parseAsString,
  queriesEqual,
  serializeDateShort,
  serializeOptionalBoolean,
  serializeString,
  serializeStringArray
} from 'sonar-ui-common/helpers/query';

export interface Query {
  activation: boolean | undefined;
  activationSeverities: string[];
  availableSince: Date | undefined;
  compareToProfile: string | undefined;
  cwe: string[];
  inheritance: T.RuleInheritance | undefined;
  languages: string[];
  owaspTop10: string[];
  profile: string | undefined;
  repositories: string[];
  ruleKey: string | undefined;
  sansTop25: string[];
  searchQuery: string | undefined;
  severities: string[];
  sonarsourceSecurity: string[];
  statuses: string[];
  tags: string[];
  template: boolean | undefined;
  types: string[];
}

export type FacetKey = keyof Query;

export interface Facet {
  [value: string]: number;
}

export type Facets = { [F in FacetKey]?: Facet };

export type OpenFacets = T.Dict<boolean>;

export interface Activation {
  inherit: T.RuleInheritance;
  severity: string;
}

export interface Actives {
  [rule: string]: {
    [profile: string]: Activation;
  };
}

export function parseQuery(query: T.RawQuery): Query {
  return {
    activation: parseAsOptionalBoolean(query.activation),
    activationSeverities: parseAsArray(query.active_severities, parseAsString),
    availableSince: parseAsDate(query.available_since),
    compareToProfile: parseAsOptionalString(query.compareToProfile),
    cwe: parseAsArray(query.cwe, parseAsString),
    inheritance: parseAsInheritance(query.inheritance),
    languages: parseAsArray(query.languages, parseAsString),
    owaspTop10: parseAsArray(query.owaspTop10, parseAsString),
    profile: parseAsOptionalString(query.qprofile),
    repositories: parseAsArray(query.repositories, parseAsString),
    ruleKey: parseAsOptionalString(query.rule_key),
    sansTop25: parseAsArray(query.sansTop25, parseAsString),
    searchQuery: parseAsOptionalString(query.q),
    severities: parseAsArray(query.severities, parseAsString),
    sonarsourceSecurity: parseAsArray(query.sonarsourceSecurity, parseAsString),
    statuses: parseAsArray(query.statuses, parseAsString),
    tags: parseAsArray(query.tags, parseAsString),
    template: parseAsOptionalBoolean(query.is_template),
    types: parseAsArray(query.types, parseAsString)
  };
}

export function serializeQuery(query: Query): T.RawQuery {
  return cleanQuery({
    activation: serializeOptionalBoolean(query.activation),
    active_severities: serializeStringArray(query.activationSeverities),
    available_since: serializeDateShort(query.availableSince),
    compareToProfile: serializeString(query.compareToProfile),
    cwe: serializeStringArray(query.cwe),
    inheritance: serializeInheritance(query.inheritance),
    is_template: serializeOptionalBoolean(query.template),
    languages: serializeStringArray(query.languages),
    owaspTop10: serializeStringArray(query.owaspTop10),
    q: serializeString(query.searchQuery),
    qprofile: serializeString(query.profile),
    repositories: serializeStringArray(query.repositories),
    rule_key: serializeString(query.ruleKey),
    sansTop25: serializeStringArray(query.sansTop25),
    severities: serializeStringArray(query.severities),
    sonarsourceSecurity: serializeStringArray(query.sonarsourceSecurity),
    statuses: serializeStringArray(query.statuses),
    tags: serializeStringArray(query.tags),
    types: serializeStringArray(query.types)
  });
}

export function areQueriesEqual(a: T.RawQuery, b: T.RawQuery) {
  return queriesEqual(parseQuery(a), parseQuery(b));
}

export function shouldRequestFacet(facet: string): facet is FacetKey {
  const facetsToRequest = [
    'activationSeverities',
    'cwe',
    'languages',
    'owaspTop10',
    'repositories',
    'sansTop25',
    'severities',
    'sonarsourceSecurity',
    'standard',
    'statuses',
    'tags',
    'types'
  ];
  return facetsToRequest.includes(facet);
}

export function getServerFacet(facet: FacetKey) {
  return facet === 'activationSeverities' ? 'active_severities' : facet;
}

export function getAppFacet(serverFacet: string): FacetKey {
  return serverFacet === 'active_severities' ? 'activationSeverities' : (serverFacet as FacetKey);
}

export function getOpen(query: T.RawQuery) {
  return query.open;
}

export function hasRuleKey(query: T.RawQuery) {
  return Boolean(query.rule_key);
}

function parseAsInheritance(value?: string): T.RuleInheritance | undefined {
  if (value === 'INHERITED' || value === 'NONE' || value === 'OVERRIDES') {
    return value;
  } else {
    return undefined;
  }
}

function serializeInheritance(value: T.RuleInheritance | undefined): string | undefined {
  return value;
}
