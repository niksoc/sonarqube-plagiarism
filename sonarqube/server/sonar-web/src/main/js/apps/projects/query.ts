/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { VISUALIZATIONS } from './utils';

type Level = 'ERROR' | 'WARN' | 'OK';

export interface Query {
  gate?: Level;
  reliability?: number;
  new_reliability?: number;
  security?: number;
  new_security?: number;
  security_review_rating?: number;
  new_security_review_rating?: number;
  maintainability?: number;
  new_maintainability?: number;
  coverage?: number;
  new_coverage?: number;
  duplications?: number;
  new_duplications?: number;
  size?: number;
  new_lines?: number;
  languages?: string[];
  tags?: string[];
  search?: string;
  sort?: string;
  view?: string;
  visualization?: string;
  [x: string]: string | number | string[] | undefined;
}

export function parseUrlQuery(urlQuery: T.RawQuery): Query {
  return {
    gate: getAsLevel(urlQuery['gate']),
    reliability: getAsNumericRating(urlQuery['reliability']),
    new_reliability: getAsNumericRating(urlQuery['new_reliability']),
    security: getAsNumericRating(urlQuery['security']),
    new_security: getAsNumericRating(urlQuery['new_security']),
    security_review_rating: getAsNumericRating(urlQuery['security_review']),
    new_security_review_rating: getAsNumericRating(urlQuery['new_security_review']),
    maintainability: getAsNumericRating(urlQuery['maintainability']),
    new_maintainability: getAsNumericRating(urlQuery['new_maintainability']),
    coverage: getAsNumericRating(urlQuery['coverage']),
    new_coverage: getAsNumericRating(urlQuery['new_coverage']),
    duplications: getAsNumericRating(urlQuery['duplications']),
    new_duplications: getAsNumericRating(urlQuery['new_duplications']),
    size: getAsNumericRating(urlQuery['size']),
    new_lines: getAsNumericRating(urlQuery['new_lines']),
    languages: getAsStringArray(urlQuery['languages']),
    tags: getAsStringArray(urlQuery['tags']),
    search: getAsString(urlQuery['search']),
    sort: getAsString(urlQuery['sort']),
    view: getView(urlQuery['view']),
    visualization: getVisualization(urlQuery['visualization'])
  };
}

export function convertToFilter(query: Query, isFavorite: boolean): string {
  const conditions: string[] = [];

  if (isFavorite) {
    conditions.push('isFavorite');
  }

  if (query['gate'] != null) {
    conditions.push(mapPropertyToMetric('gate') + ' = ' + query['gate']);
  }

  ['coverage', 'new_coverage'].forEach(property =>
    pushMetricToArray(query, property, conditions, convertCoverage)
  );

  ['duplications', 'new_duplications'].forEach(property =>
    pushMetricToArray(query, property, conditions, convertDuplications)
  );

  ['size', 'new_lines'].forEach(property =>
    pushMetricToArray(query, property, conditions, convertSize)
  );

  [
    'reliability',
    'security',
    'security_review_rating',
    'maintainability',
    'new_reliability',
    'new_security',
    'new_security_review_rating',
    'new_maintainability'
  ].forEach(property => pushMetricToArray(query, property, conditions, convertIssuesRating));

  ['languages', 'tags'].forEach(property =>
    pushMetricToArray(query, property, conditions, convertArrayMetric)
  );

  if (query['search'] != null) {
    conditions.push(`${mapPropertyToMetric('search')} = "${query['search']}"`);
  }

  return conditions.join(' and ');
}

const visualizationParams = ['sort', 'view', 'visualization'];

export function hasFilterParams(query: Query) {
  return Object.keys(query)
    .filter(key => !visualizationParams.includes(key))
    .some(key => query[key] !== undefined);
}

export function hasVisualizationParams(query: Query) {
  return Object.keys(query)
    .filter(key => visualizationParams.includes(key))
    .some(key => query[key] !== undefined);
}

function getAsNumericRating(value: any): number | undefined {
  if (value === '' || value == null || isNaN(value)) {
    return undefined;
  }
  const num = Number(value);
  return num > 0 && num < 7 ? num : undefined;
}

function getAsLevel(value: any): Level | undefined {
  if (value === 'ERROR' || value === 'WARN' || value === 'OK') {
    return value;
  }
  return undefined;
}

function getAsString(value: any): string | undefined {
  if (typeof value !== 'string' || !value) {
    return undefined;
  }
  return value;
}

function getAsStringArray(value: any): string[] | undefined {
  if (typeof value !== 'string' || !value) {
    return undefined;
  }
  return value.split(',');
}

function getView(value: any): string | undefined {
  return typeof value !== 'string' || value === 'overall' ? undefined : value;
}

function getVisualization(value: string): string | undefined {
  return VISUALIZATIONS.includes(value) ? value : undefined;
}

function convertIssuesRating(metric: string, rating: number): string {
  if (rating > 1 && rating < 5) {
    return `${metric} >= ${rating}`;
  } else {
    return `${metric} = ${rating}`;
  }
}

function convertCoverage(metric: string, coverage: number): string {
  switch (coverage) {
    case 1:
      return metric + ' >= 80';
    case 2:
      return metric + ' < 80';
    case 3:
      return metric + ' < 70';
    case 4:
      return metric + ' < 50';
    case 5:
      return metric + ' < 30';
    case 6:
      return metric + '= NO_DATA';
    default:
      return '';
  }
}

function convertDuplications(metric: string, duplications: number): string {
  switch (duplications) {
    case 1:
      return metric + ' < 3';
    case 2:
      return metric + ' >= 3';
    case 3:
      return metric + ' >= 5';
    case 4:
      return metric + ' >= 10';
    case 5:
      return metric + ' >= 20';
    case 6:
      return metric + '= NO_DATA';
    default:
      return '';
  }
}

function convertSize(metric: string, size: number): string {
  switch (size) {
    case 1:
      return metric + ' < 1000';
    case 2:
      return metric + ' >= 1000';
    case 3:
      return metric + ' >= 10000';
    case 4:
      return metric + ' >= 100000';
    case 5:
      return metric + ' >= 500000';
    default:
      return '';
  }
}

function mapPropertyToMetric(property?: string): string | undefined {
  const map: T.Dict<string> = {
    analysis_date: 'analysisDate',
    reliability: 'reliability_rating',
    new_reliability: 'new_reliability_rating',
    security: 'security_rating',
    new_security: 'new_security_rating',
    security_review_rating: 'security_review_rating',
    new_security_review_rating: 'new_security_review_rating',
    maintainability: 'sqale_rating',
    new_maintainability: 'new_maintainability_rating',
    coverage: 'coverage',
    new_coverage: 'new_coverage',
    duplications: 'duplicated_lines_density',
    new_duplications: 'new_duplicated_lines_density',
    size: 'ncloc',
    new_lines: 'new_lines',
    gate: 'alert_status',
    languages: 'languages',
    tags: 'tags',
    search: 'query'
  };
  return property && map[property];
}

function pushMetricToArray(
  query: Query,
  property: string,
  conditionsArray: string[],
  convertFunction: (metric: string, value: any) => string
): void {
  const metric = mapPropertyToMetric(property);
  if (query[property] != null && metric) {
    conditionsArray.push(convertFunction(metric, query[property]));
  }
}

function convertArrayMetric(metric: string, items: string | string[]): string {
  if (!Array.isArray(items) || items.length < 2) {
    return metric + ' = ' + items;
  }
  return `${metric} IN (${items.join(', ')})`;
}
