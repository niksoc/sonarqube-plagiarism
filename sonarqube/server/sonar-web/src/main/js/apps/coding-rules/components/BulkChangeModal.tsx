/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { ResetButtonLink, SubmitButton } from 'sonar-ui-common/components/controls/buttons';
import Modal from 'sonar-ui-common/components/controls/Modal';
import Select from 'sonar-ui-common/components/controls/Select';
import { Alert } from 'sonar-ui-common/components/ui/Alert';
import { translate, translateWithParameters } from 'sonar-ui-common/helpers/l10n';
import { formatMeasure } from 'sonar-ui-common/helpers/measures';
import { bulkActivateRules, bulkDeactivateRules, Profile } from '../../../api/quality-profiles';
import { Query, serializeQuery } from '../query';

interface Props {
  action: string;
  languages: T.Languages;
  onClose: () => void;
  organization: string | undefined;
  profile?: Profile;
  query: Query;
  referencedProfiles: T.Dict<Profile>;
  total: number;
}

interface ActivationResult {
  failed: number;
  profile: string;
  succeeded: number;
}

interface State {
  finished: boolean;
  results: ActivationResult[];
  selectedProfiles: any[];
  submitting: boolean;
}

export default class BulkChangeModal extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);

    // if there is only one possible option for profile, select it immediately
    const selectedProfiles = [];
    const availableProfiles = this.getAvailableQualityProfiles(props);
    if (availableProfiles.length === 1) {
      selectedProfiles.push(availableProfiles[0].key);
    }

    this.state = { finished: false, results: [], selectedProfiles, submitting: false };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleProfileSelect = (options: { value: string }[]) => {
    const selectedProfiles = options.map(option => option.value);
    this.setState({ selectedProfiles });
  };

  getAvailableQualityProfiles = ({ query, referencedProfiles } = this.props) => {
    let profiles = Object.values(referencedProfiles);
    if (query.languages.length > 0) {
      profiles = profiles.filter(profile => query.languages.includes(profile.language));
    }
    return profiles
      .filter(profile => profile.actions && profile.actions.edit)
      .filter(profile => !profile.isBuiltIn);
  };

  processResponse = (profile: string, response: any) => {
    if (this.mounted) {
      const result: ActivationResult = {
        failed: response.failed || 0,
        profile,
        succeeded: response.succeeded || 0
      };
      this.setState(state => ({ results: [...state.results, result] }));
    }
  };

  sendRequests = () => {
    let looper = Promise.resolve();

    // serialize the query, but delete the `profile`
    const data = serializeQuery(this.props.query);
    delete data.profile;

    const method = this.props.action === 'activate' ? bulkActivateRules : bulkDeactivateRules;

    // if a profile is selected in the facet, pick it
    // otherwise take all profiles selected in the dropdown
    const profiles: string[] = this.props.profile
      ? [this.props.profile.key]
      : this.state.selectedProfiles;

    for (const profile of profiles) {
      looper = looper.then(() =>
        method({
          ...data,
          organization: this.props.organization,
          targetKey: profile
        }).then(response => this.processResponse(profile, response))
      );
    }
    return looper;
  };

  handleFormSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.setState({ submitting: true });
    this.sendRequests().then(
      () => {
        if (this.mounted) {
          this.setState({ finished: true, submitting: false });
        }
      },
      () => {
        if (this.mounted) {
          this.setState({ submitting: false });
        }
      }
    );
  };

  renderResult = (result: ActivationResult) => {
    const { profile: profileKey } = result;
    const profile = this.props.referencedProfiles[profileKey];
    if (!profile) {
      return null;
    }
    const { languages } = this.props;
    const language = languages[profile.language]
      ? languages[profile.language].name
      : profile.language;
    return (
      <Alert key={result.profile} variant={result.failed === 0 ? 'success' : 'warning'}>
        {result.failed
          ? translateWithParameters(
              'coding_rules.bulk_change.warning',
              profile.name,
              language,
              result.succeeded,
              result.failed
            )
          : translateWithParameters(
              'coding_rules.bulk_change.success',
              profile.name,
              language,
              result.succeeded
            )}
      </Alert>
    );
  };

  renderProfileSelect = () => {
    const profiles = this.getAvailableQualityProfiles();
    const options = profiles.map(profile => ({
      label: `${profile.name} - ${profile.languageName}`,
      value: profile.key
    }));
    return (
      <Select
        multi={true}
        onChange={this.handleProfileSelect}
        options={options}
        value={this.state.selectedProfiles}
      />
    );
  };

  render() {
    const { action, profile, total } = this.props;
    const header =
      action === 'activate'
        ? `${translate('coding_rules.activate_in_quality_profile')} (${formatMeasure(
            total,
            'INT'
          )} ${translate('coding_rules._rules')})`
        : `${translate('coding_rules.deactivate_in_quality_profile')} (${formatMeasure(
            total,
            'INT'
          )} ${translate('coding_rules._rules')})`;

    return (
      <Modal contentLabel={header} onRequestClose={this.props.onClose} size="small">
        <form onSubmit={this.handleFormSubmit}>
          <header className="modal-head">
            <h2>{header}</h2>
          </header>

          <div className="modal-body">
            {this.state.results.map(this.renderResult)}

            {!this.state.finished && !this.state.submitting && (
              <div className="modal-field">
                <h3>
                  <label htmlFor="coding-rules-bulk-change-profile">
                    {action === 'activate'
                      ? translate('coding_rules.activate_in')
                      : translate('coding_rules.deactivate_in')}
                  </label>
                </h3>
                {profile ? (
                  <span>
                    {profile.name}
                    {' â€” '}
                    {translate('are_you_sure')}
                  </span>
                ) : (
                  this.renderProfileSelect()
                )}
              </div>
            )}
          </div>

          <footer className="modal-foot">
            {this.state.submitting && <i className="spinner spacer-right" />}
            {!this.state.finished && (
              <SubmitButton disabled={this.state.submitting} id="coding-rules-submit-bulk-change">
                {translate('apply')}
              </SubmitButton>
            )}
            <ResetButtonLink onClick={this.props.onClose}>
              {this.state.finished ? translate('close') : translate('cancel')}
            </ResetButtonLink>
          </footer>
        </form>
      </Modal>
    );
  }
}
