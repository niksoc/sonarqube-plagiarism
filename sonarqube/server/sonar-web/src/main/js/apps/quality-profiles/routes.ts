/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { lazyLoad } from 'sonar-ui-common/components/lazyLoad';

const routes = [
  {
    component: lazyLoad(() => import('./components/AppContainer')),
    indexRoute: { component: lazyLoad(() => import('./home/HomeContainer')) },
    childRoutes: [
      {
        component: lazyLoad(() => import('./components/ProfileContainer')),
        childRoutes: [
          {
            path: 'show',
            component: lazyLoad(() => import('./details/ProfileDetails'))
          },
          {
            path: 'changelog',
            component: lazyLoad(() => import('./changelog/ChangelogContainer'))
          },
          {
            path: 'compare',
            component: lazyLoad(() => import('./compare/ComparisonContainer'))
          }
        ]
      }
    ]
  }
];

export default routes;
