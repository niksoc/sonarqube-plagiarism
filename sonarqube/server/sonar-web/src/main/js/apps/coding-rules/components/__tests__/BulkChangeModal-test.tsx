/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { submit, waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import { bulkActivateRules, bulkDeactivateRules } from '../../../../api/quality-profiles';
import { mockLanguage, mockQualityProfile } from '../../../../helpers/testMocks';
import { Query } from '../../query';
import BulkChangeModal from '../BulkChangeModal';

jest.mock('../../../../api/quality-profiles', () => ({
  bulkActivateRules: jest.fn().mockResolvedValue({ failed: 0, succeeded: 2 }),
  bulkDeactivateRules: jest.fn().mockResolvedValue({ failed: 2, succeeded: 0 })
}));

beforeEach(jest.clearAllMocks);

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(shallowRender({ profile: undefined })).toMatchSnapshot('no profile pre-selected');
  expect(shallowRender({ action: 'deactivate' })).toMatchSnapshot('deactivate action');
  expect(
    shallowRender().setState({
      results: [
        { failed: 2, profile: 'foo', succeeded: 0 },
        { failed: 0, profile: 'bar', succeeded: 2 }
      ]
    })
  ).toMatchSnapshot('results');
  expect(shallowRender().setState({ submitting: true })).toMatchSnapshot('submitting');
  expect(shallowRender().setState({ finished: true })).toMatchSnapshot('finished');
});

it('should pre-select a profile if only 1 is available', () => {
  const profile = mockQualityProfile({
    actions: { edit: true },
    isBuiltIn: false,
    key: 'foo',
    language: 'js'
  });
  const wrapper = shallowRender({ profile: undefined, referencedProfiles: { foo: profile } });
  expect(wrapper.state().selectedProfiles).toEqual(['foo']);
});

it('should handle profile selection', () => {
  const wrapper = shallowRender();
  wrapper.instance().handleProfileSelect([{ value: 'foo' }, { value: 'bar' }]);
  expect(wrapper.state().selectedProfiles).toEqual(['foo', 'bar']);
});

it('should handle form submission', async () => {
  const wrapper = shallowRender({ profile: undefined });
  wrapper.setState({ selectedProfiles: ['foo', 'bar'] });

  // Activate.
  submit(wrapper.find('form'));
  await waitAndUpdate(wrapper);
  expect(bulkActivateRules).toBeCalledWith(expect.objectContaining({ targetKey: 'foo' }));

  await waitAndUpdate(wrapper);
  expect(bulkActivateRules).toBeCalledWith(expect.objectContaining({ targetKey: 'bar' }));

  await waitAndUpdate(wrapper);
  expect(wrapper.state().results).toEqual([
    { failed: 0, profile: 'foo', succeeded: 2 },
    { failed: 0, profile: 'bar', succeeded: 2 }
  ]);

  // Deactivate.
  wrapper.setProps({ action: 'deactivate' }).setState({ results: [] });
  submit(wrapper.find('form'));
  await waitAndUpdate(wrapper);
  expect(bulkDeactivateRules).toBeCalledWith(expect.objectContaining({ targetKey: 'foo' }));

  await waitAndUpdate(wrapper);
  expect(bulkDeactivateRules).toBeCalledWith(expect.objectContaining({ targetKey: 'bar' }));

  await waitAndUpdate(wrapper);
  expect(wrapper.state().results).toEqual([
    { failed: 2, profile: 'foo', succeeded: 0 },
    { failed: 2, profile: 'bar', succeeded: 0 }
  ]);
});

function shallowRender(props: Partial<BulkChangeModal['props']> = {}) {
  return shallow<BulkChangeModal>(
    <BulkChangeModal
      action="activate"
      languages={{ js: mockLanguage() }}
      onClose={jest.fn()}
      organization={undefined}
      profile={mockQualityProfile()}
      query={{ languages: ['js'] } as Query}
      referencedProfiles={{
        foo: mockQualityProfile({ key: 'foo' }),
        bar: mockQualityProfile({ key: 'bar' })
      }}
      total={42}
      {...props}
    />
  );
}
