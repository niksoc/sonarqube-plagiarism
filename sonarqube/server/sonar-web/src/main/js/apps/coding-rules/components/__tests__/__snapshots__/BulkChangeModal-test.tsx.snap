// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: deactivate action 1`] = `
<Modal
  contentLabel="coding_rules.deactivate_in_quality_profile (42 coding_rules._rules)"
  onRequestClose={[MockFunction]}
  size="small"
>
  <form
    onSubmit={[Function]}
  >
    <header
      className="modal-head"
    >
      <h2>
        coding_rules.deactivate_in_quality_profile (42 coding_rules._rules)
      </h2>
    </header>
    <div
      className="modal-body"
    >
      <div
        className="modal-field"
      >
        <h3>
          <label
            htmlFor="coding-rules-bulk-change-profile"
          >
            coding_rules.deactivate_in
          </label>
        </h3>
        <span>
          name
           — 
          are_you_sure
        </span>
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <SubmitButton
        disabled={false}
        id="coding-rules-submit-bulk-change"
      >
        apply
      </SubmitButton>
      <ResetButtonLink
        onClick={[MockFunction]}
      >
        cancel
      </ResetButtonLink>
    </footer>
  </form>
</Modal>
`;

exports[`should render correctly: default 1`] = `
<Modal
  contentLabel="coding_rules.activate_in_quality_profile (42 coding_rules._rules)"
  onRequestClose={[MockFunction]}
  size="small"
>
  <form
    onSubmit={[Function]}
  >
    <header
      className="modal-head"
    >
      <h2>
        coding_rules.activate_in_quality_profile (42 coding_rules._rules)
      </h2>
    </header>
    <div
      className="modal-body"
    >
      <div
        className="modal-field"
      >
        <h3>
          <label
            htmlFor="coding-rules-bulk-change-profile"
          >
            coding_rules.activate_in
          </label>
        </h3>
        <span>
          name
           — 
          are_you_sure
        </span>
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <SubmitButton
        disabled={false}
        id="coding-rules-submit-bulk-change"
      >
        apply
      </SubmitButton>
      <ResetButtonLink
        onClick={[MockFunction]}
      >
        cancel
      </ResetButtonLink>
    </footer>
  </form>
</Modal>
`;

exports[`should render correctly: finished 1`] = `
<Modal
  contentLabel="coding_rules.activate_in_quality_profile (42 coding_rules._rules)"
  onRequestClose={[MockFunction]}
  size="small"
>
  <form
    onSubmit={[Function]}
  >
    <header
      className="modal-head"
    >
      <h2>
        coding_rules.activate_in_quality_profile (42 coding_rules._rules)
      </h2>
    </header>
    <div
      className="modal-body"
    />
    <footer
      className="modal-foot"
    >
      <ResetButtonLink
        onClick={[MockFunction]}
      >
        close
      </ResetButtonLink>
    </footer>
  </form>
</Modal>
`;

exports[`should render correctly: no profile pre-selected 1`] = `
<Modal
  contentLabel="coding_rules.activate_in_quality_profile (42 coding_rules._rules)"
  onRequestClose={[MockFunction]}
  size="small"
>
  <form
    onSubmit={[Function]}
  >
    <header
      className="modal-head"
    >
      <h2>
        coding_rules.activate_in_quality_profile (42 coding_rules._rules)
      </h2>
    </header>
    <div
      className="modal-body"
    >
      <div
        className="modal-field"
      >
        <h3>
          <label
            htmlFor="coding-rules-bulk-change-profile"
          >
            coding_rules.activate_in
          </label>
        </h3>
        <Select
          multi={true}
          onChange={[Function]}
          options={Array []}
          value={Array []}
        />
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <SubmitButton
        disabled={false}
        id="coding-rules-submit-bulk-change"
      >
        apply
      </SubmitButton>
      <ResetButtonLink
        onClick={[MockFunction]}
      >
        cancel
      </ResetButtonLink>
    </footer>
  </form>
</Modal>
`;

exports[`should render correctly: results 1`] = `
<Modal
  contentLabel="coding_rules.activate_in_quality_profile (42 coding_rules._rules)"
  onRequestClose={[MockFunction]}
  size="small"
>
  <form
    onSubmit={[Function]}
  >
    <header
      className="modal-head"
    >
      <h2>
        coding_rules.activate_in_quality_profile (42 coding_rules._rules)
      </h2>
    </header>
    <div
      className="modal-body"
    >
      <Alert
        key="foo"
        variant="warning"
      >
        coding_rules.bulk_change.warning.name.CSS.0.2
      </Alert>
      <Alert
        key="bar"
        variant="success"
      >
        coding_rules.bulk_change.success.name.CSS.2
      </Alert>
      <div
        className="modal-field"
      >
        <h3>
          <label
            htmlFor="coding-rules-bulk-change-profile"
          >
            coding_rules.activate_in
          </label>
        </h3>
        <span>
          name
           — 
          are_you_sure
        </span>
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <SubmitButton
        disabled={false}
        id="coding-rules-submit-bulk-change"
      >
        apply
      </SubmitButton>
      <ResetButtonLink
        onClick={[MockFunction]}
      >
        cancel
      </ResetButtonLink>
    </footer>
  </form>
</Modal>
`;

exports[`should render correctly: submitting 1`] = `
<Modal
  contentLabel="coding_rules.activate_in_quality_profile (42 coding_rules._rules)"
  onRequestClose={[MockFunction]}
  size="small"
>
  <form
    onSubmit={[Function]}
  >
    <header
      className="modal-head"
    >
      <h2>
        coding_rules.activate_in_quality_profile (42 coding_rules._rules)
      </h2>
    </header>
    <div
      className="modal-body"
    />
    <footer
      className="modal-foot"
    >
      <i
        className="spinner spacer-right"
      />
      <SubmitButton
        disabled={true}
        id="coding-rules-submit-bulk-change"
      >
        apply
      </SubmitButton>
      <ResetButtonLink
        onClick={[MockFunction]}
      >
        cancel
      </ResetButtonLink>
    </footer>
  </form>
</Modal>
`;
